include/aws/crt/http/HttpConnection.h

```diff
@@ -0,0 +1,356 @@
+
+            /**
+             * POD structure used for setting up an Http Request
+             */
+            struct HttpRequestOptions
+            {
+                /**
+                 * Http verb to use (e.g. GET, POST, PUT, DELETE, HEAD....). If you are using a custom verb, that
+                 * can be set here as well.
+                 *
+                 * this value is copied internally.
+                 */
+                ByteCursor method;
+                /**
+                 * Usually the Path and Query portion of the request uri (assuming the host header has been set).
+                 * There's not validation on this, it can also be the absolute uri if you want/need that.
+                 *
+                 * A helpful utility for setting this value can be found in Aws::Crt::Io::Uri
+                 *
+                 * this value is copied internally.
+                 */
+                ByteCursor uri;
+                /**
+                 * Array of headers to use for the Request. These values will be copied internally.
+                 */
+                aws_http_header *headerArray;
+                /**
+                 * Length of `headerArray`.
+                 */
+                std::size_t headerArrayLength;
+                /**
+                 * See `OnConnectionShutdown` for more info. This value can be unset if you don't need to send a body.
+                 */
+                OnStreamOutgoingBody onStreamOutgoingBody;
+                /**
+                 * See `OnIncomingHeaders` for more info. This value must be set.
+                 */
+                OnIncomingHeaders onIncomingHeaders;
+                OnIncomingHeadersBlockDone onIncomingHeadersBlockDone;
+                /**
+                 * See `OnIncomingBody` for more info. This value can be unset if you will not be receiving a body.
+                 */
+                OnIncomingBody onIncomingBody;
+                /**
+                 * See `OnStreamComplete` for more info. This value can be unset.
+                 */
+                OnStreamComplete onStreamComplete;
+            };
+
+            /**
+             * Represents a single http message exchange (request/response) or in H2, it can also represent
+             * a PUSH_PROMISE followed by the accompanying Response.
+             */
+            class AWS_CRT_CPP_API HttpStream final
+            {
+              public:
+                ~HttpStream();
+                HttpStream(const HttpStream &) = delete;
+                HttpStream(HttpStream &&) = delete;
+                HttpStream &operator=(const HttpStream &) = delete;
+                HttpStream &operator=(HttpStream &&) = delete;
+
+                /**
+                 * Get the underlying connection for the stream.
+                 */
+                const std::shared_ptr<HttpConnection> &GetConnection() const noexcept;
```

Comment:
Is it invariant that an HttpStream always have a connection? if yes, then return by reference instead of shared_ptr